session/v2/token.go

```diff
@@ -0,0 +1,1028 @@
+	l.exp = exp
+}
+
+// ValidAt checks if the lifetime is valid at the given timestamp.
+func (l Lifetime) ValidAt(timestamp uint64) bool {
+	return timestamp >= l.iat && timestamp <= l.exp && timestamp >= l.nbf
+}
+
+// Target represents an account identifier in session token V2.
+// It can be either a direct OwnerID reference or an NNS name.
+type Target struct {
+	ownerID user.ID
+	nnsName string
+}
+
+// NewTargetUser creates a new Target from user.ID.
+func NewTargetUser(id user.ID) Target {
+	return Target{ownerID: id}
+}
+
+// NewTargetNamed creates a new Target from NNS name.
+func NewTargetNamed(nnsName string) Target {
+	return Target{nnsName: nnsName}
+}
+
+// IsOwnerID returns true if target uses direct OwnerID reference.
+func (t Target) IsOwnerID() bool {
+	return !t.ownerID.IsZero()
+}
+
+// IsNNS returns true if target uses NNS name reference.
+func (t Target) IsNNS() bool {
+	return t.nnsName != ""
+}
+
+// OwnerID returns the owner ID if target uses direct reference.
+func (t Target) OwnerID() user.ID {
+	return t.ownerID
+}
+
+// NNSName returns the NNS name if target uses NNS reference.
+func (t Target) NNSName() string {
+	return t.nnsName
+}
+
+// Equals checks if two targets are equal.
+func (t Target) Equals(other Target) bool {
+	return t.ownerID == other.ownerID && t.nnsName == other.nnsName
+}
+
+// String returns string representation of the target.
+func (t Target) String() string {
+	if t.IsOwnerID() {
+		return t.ownerID.String()
+	}
+	return t.nnsName
+}
+
+// IsEmpty checks if the target is empty.
+func (t Target) IsEmpty() bool {
+	return t.ownerID.IsZero() && t.nnsName == ""
+}
+
+func (t Target) protoMessage() *protosession.Target {
+	switch {
+	case t.IsOwnerID():
+		return &protosession.Target{
+			Identifier: &protosession.Target_OwnerId{
+				OwnerId: t.ownerID.ProtoMessage(),
+			},
+		}
+	case t.IsNNS():
+		return &protosession.Target{
+			Identifier: &protosession.Target_NnsName{
+				NnsName: t.nnsName,
+			},
+		}
+	default:
+		return nil
+	}
+}
+
+func (t *Target) fromProtoMessage(m *protosession.Target) error {
+	if m == nil {
+		return errors.New("nil target")
+	}
+
+	switch id := m.Identifier.(type) {
+	case *protosession.Target_OwnerId:
+		if id.OwnerId == nil {
+			return errors.New("nil owner ID in target")
+		}
+		return t.ownerID.FromProtoMessage(id.OwnerId)
+	case *protosession.Target_NnsName:
+		if id.NnsName == "" {
+			return errors.New("empty NNS name in target")
+		}
+		t.nnsName = id.NnsName
+		return nil
+	default:
+		return fmt.Errorf("unknown target identifier type: %T", id)
+	}
+}
+
+// Context represents a unified session context for both object and container operations.
+type Context struct {
+	container cid.ID
+	objects   []oid.ID
+	verbs     []Verb
+}
+
+// NewContext creates a new unified session context.
+// Returns error if the number of verbs exceeds MaxVerbsPerContext.
+func NewContext(container cid.ID, verbs []Verb) (Context, error) {
```

Comment:
are empty verbs valid?

Reply:
BTW, treating empty verbs as wildcard can be a useful extension, but not immediately.